src/lib/Inference.hs

```diff
@@ -2156,30 +2200,42 @@ synthTopBlock block = do
   (liftExcept =<<) $ liftDictSynthTraverserM $ traverseGenericE block
 {-# SCC synthTopBlock #-}
 
--- main entrypoint to dictionary synthesizer
-trySynthDict :: (Emits n, Builder m, Fallible1 m, EnvExtender m, EnvReader m)
-             => Type n -> m n (Atom n)
-trySynthDict ty = trySynthDictBlock ty >>= emitBlock
+synthInstanceDef
+  :: (EnvReader m, Fallible1 m) => InstanceDef n -> m n (InstanceDef n)
+synthInstanceDef (InstanceDef className bs params body) = do
+  liftExceptEnvReaderM $ refreshAbs (Abs bs (ListE params `PairE` body))
+    \bs' (ListE params' `PairE` InstanceBody superclasses methods) -> do
+      methods' <- mapM synthTopBlock methods
+      return $ InstanceDef className bs' params' $ InstanceBody superclasses methods'
 
-trySynthDictBlock :: (Fallible1 m, EnvReader m) => Type n -> m n (Block n)
-trySynthDictBlock ty = do
+-- main entrypoint to dictionary synthesizer
+trySynthTerm :: (Fallible1 m, EnvReader m) => Type n -> m n (Atom n)
+trySynthTerm ty = do
   hasInferenceVars ty >>= \case
     True -> throw TypeErr "Can't synthesize a dictionary for a type with inference vars"
     False -> do
-      solutions <- liftSyntherM $ buildBlock do
-                     ty' <- sinkM ty
-                     synthDict ty'
+      synthTy <- typeAsSynthType ty
+      solutions <- liftSyntherM $ synthTerm synthTy
       let solutionsInstance = filter ((== UsedInstance) . snd) solutions
       unless (length solutionsInstance <= 1) $
         throw TypeErr $ "Multiple candidate class dictionaries for: " ++ pprint ty
       case solutions of
         [] -> throw TypeErr $ "Couldn't synthesize a class dictionary for: " ++ pprint ty
-        (d, _):_ -> return d
-{-# SCC trySynthDictBlock #-}
+        (d, _):_ -> cheapNormalize d
+{-# SCC trySynthTerm #-}
 
-data Givens n = Givens { fromGivens :: HM.HashMap (EKey Type n) (Given n) }
+-- The type of a `SynthExpr` atom must be `DictTy _` or `Pi _`. If it's `Pi _`,
+-- then the atom itself is either a variable or a trivial lambda whose body
+-- is also a `SynthExpr`.
+type SynthExpr = Atom
 
-type Given = Block
+data SynthType n =
+   SynthDictType (DictType n)
+ | SynthPiType (Abs PiBinder SynthType n)
+   deriving (Show, Generic)
```

Comment:
What's the benefit of maintaining `SynthType`? It seems like we have to do conversions in quite a few places and we have to write whole functions that e.g. interpret an instance as a `SynthType`, even though we already wrote a function that creates a `Type` for an instance!

Reply:
Good point. I added it because I needed help figuring out which types were valid arguments to `instantiateSynthArgs`. But it would probably be fine to just make `SynthType` an alias with a comment explaining that it has to be either `DictTy` or `Pi`.

Reply:
Yeah I think that would be a bit nicer and more consistent with what we do in other places.

Reply:
On reflection, I think I prefer it the way it is currently. It's more precise and not much extra code. There's only one conversion, `typeAsSynthType`, and that's actually a meaningful function, not just boilerplate, because it checks whether the type is the sort of thing we know how to synthesize.